go/otgrpc/client.go

```diff
@@ -81,3 +72,168 @@ func OpenTracingClientInterceptor(tracer opentracing.Tracer, optFuncs ...Option)
 		return err
 	}
 }
+
+// OpenTracingStreamClientInterceptor returns a grpc.StreamClientInterceptor suitable
+// for use in a grpc.Dial call. The interceptor instruments streaming RPCs by creating
+// a single span to correspond to the lifetime of the RPC's stream.
+//
+// For example:
+//
+//     conn, err := grpc.Dial(
+//         address,
+//         ...,  // (existing DialOptions)
+//         grpc.WithStreamInterceptor(otgrpc.OpenTracingStreamClientInterceptor(tracer)))
+//
+// All gRPC client spans will inject the OpenTracing SpanContext into the gRPC
+// metadata; they will also look in the context.Context for an active
+// in-process parent Span and establish a ChildOf reference if such a parent
+// Span could be found.
+func OpenTracingStreamClientInterceptor(tracer opentracing.Tracer, optFuncs ...Option) grpc.StreamClientInterceptor {
+	otgrpcOpts := newOptions()
+	otgrpcOpts.apply(optFuncs...)
+	return func(
+		ctx context.Context,
+		desc *grpc.StreamDesc,
+		cc *grpc.ClientConn,
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (grpc.ClientStream, error) {
+		var err error
+		var parentCtx opentracing.SpanContext
+		if parent := opentracing.SpanFromContext(ctx); parent != nil {
+			parentCtx = parent.Context()
+		}
+		if otgrpcOpts.inclusionFunc != nil &&
+			!otgrpcOpts.inclusionFunc(parentCtx, method, nil, nil) {
+			return streamer(ctx, desc, cc, method, opts...)
+		}
+
+		clientSpan := tracer.StartSpan(
+			method,
+			opentracing.ChildOf(parentCtx),
+			ext.SpanKindRPCClient,
+			gRPCComponentTag,
+		)
+		ctx = injectSpanContext(ctx, tracer, clientSpan)
+		cs, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			clientSpan.LogFields(log.String("event", "error"), log.String("message", err.Error()))
+			SetSpanTags(clientSpan, err, true)
+			clientSpan.Finish()
+			return cs, err
+		}
+		return newOpenTracingClientStream(cs, method, desc, clientSpan, otgrpcOpts), nil
+	}
+}
+
+func newOpenTracingClientStream(cs grpc.ClientStream, method string, desc *grpc.StreamDesc, clientSpan opentracing.Span, otgrpcOpts *options) grpc.ClientStream {
+	finishChan := make(chan struct{})
+
+	isFinished := new(int32)
+	*isFinished = 0
+	finishFunc := func(err error) {
+		// The current OpenTracing specification forbids finishing a span more than
+		// once. Since we have multiple code paths that could concurrently call
+		// `finishFunc`, we need to add some sort of synchronization to guard against
+		// multiple finishing.
+		if !atomic.CompareAndSwapInt32(isFinished, 0, 1) {
```

Comment:
why can't this all use a `sync.Once`? The case of `finishedChan` being closed without `<-cs.Context().Done()` being closed is exactly the case where the finalizer has run, so there's no need for this closure to know about `finishedChan` at all, and there's no need for this atomic (use sync.Once instead). What do you think?

Reply:
Not sure I follow: The finalizer calls this same closure. If the closure doesn't know about `finishedChan`, how does the channel get closed and the Go-routine not leaked?

Reply:
Just close it right in the finalizer, it's the only place that needs to bother.

Reply:
I'm open to doing that. But I'm not sure that makes the code any simpler: it's just moving the closing from one closure to another, no?

And just to be clear on the finalizer, it's only meant to handle the special case when the stream isn't read until [`io.EOF`](https://stackoverflow.com/questions/42915337/are-you-required-to-call-recv-until-you-get-io-eof-when-interacting-with-grpc-cl/42939914#42939914). I would expect the most common usage to be that `<- cs.Context().Done()` isn't closed and the finished is done by intercepting an [`io.EOF`](https://github.com/rnburn/grpc-opentracing-go/blob/master/go/otgrpc/client.go#L204).

Reply:
Ah, I see! I must have misread the first time through; apologies. I still think you can substitute the atomic for a `sync.Once` to get slightly more expected (to me) behaviour where no calls to `finishFunc` will return until the finishing is actually complete.

Reply:
I could see how that behavior could be preferred in most situations, but is it really the best fit for tracing?

Given that the tracing code should be orthogonal to the rest of the functioning of an app, I would think you'd want it to get out of the way as much as possible. For example, if one thread cancels the RPC call via the context while another thread reads an `io.EOF` from `RecvMsg`, do you really want one of the code paths to take longer just to wait on the span's finishing code to complete?

Reply:
I could argue either way, so I won't.